### ISSUE TITLE
(Windows) error: Video support is permanently disabled in this executable.

### ISSUE THREAD

**Issue by ericoporto** [ImageOptim/gifski#226]
The release binaries doesn't support video input, but in the same page it says video as input should be possible.

**Comment by kornelski** [comment]
Building ffmpeg on Windows is a pain. I've managed to get it working on Linux and macOS, but I have no hope of fixing it on Windows. You'll have to use a separate command to make PNG frames.

**Comment by camjac251** [comment]
Would these help @kornelski https://www.gyan.dev/ffmpeg/builds/

**Comment by kornelski** [comment]
No, these are builds of `ffmpeg` command-line tool, which isn't for integrating into other programs.

**Comment by andylizi** [comment]
It does provided a "shared" build for dynamic linking. I have successfully built gifski with it before, and I remember it being a relatively painless process — just followed the instruction [here](https://github.com/zmwangx/rust-ffmpeg/wiki/Notes-on-building#msvc-toolchain). `gyan.dev` is also what `ffmpeg-sys-next` used for it's [CI build test](https://github.com/zmwangx/rust-ffmpeg/blob/v4.3.8/.github/workflows/build.yml#L75).

**Comment by kornelski** [comment]
ah, ok

**Comment by kornelski** [comment]
I've tried gyan.dev/ffmpeg/builds. Only ffmpeg-4.4.1-full_build-shared.7z contained libraries. Even though they were supposedly static .lib files, compiling the ffmpeg crate with them made gifski.exe depend on these DLLs:

    avcodec-58.dll
    avformat-58.dll
    avutil-56.dll
    swscale-5.dll
    swresample-3.dll
    avfilter-7.dll

and it crashes without them (exit code: 0xc0000135, STATUS_DLL_NOT_FOUND).

To repro:

1. Install LLVM
2. Extract ffmpeg-4.4.1-full_build-shared.7z to C:\whatever\ffmpegbuild
3. `set FFMPEG_DIR=C:\whatever\ffmpegbuild`
4. `cargo run --features=video-prebuilt-static`

**Comment by andylizi** [comment]
Ah yes…… That's the thing. The "static .lib files" actually just reexport from the DLLs, and do not contain the code itself (their sizes are too small for that). I've spent a long time trying to do the same thing and eventually decided it's not possible.

I had to build my own ffmpeg in the end, which was fun.

**Comment by jalba69** [comment]
Hello,
can you give a little more details about how to have it even with the FFMPEG dll's needed please ?

I tried it but i end up with an error : 
```
Compiling imagequant v4.0.0
Compiling ffmpeg-sys-next v4.4.0 (https://github.com/kornelski/rust-ffmpeg-sys-1?rev=78458039d5fac99354f9cb078869f3be3f3af5fb#78458039)
error: failed to run custom build command for ffmpeg-sys-next v4.4.0 (https://github.com/kornelski/rust-ffmpeg-sys-1?rev=78458039d5fac99354f9cb078869f3be3f3af5fb#78458039)

Caused by:
  process didn't exit successfully: H:\gifski-main\target\debug\build\ffmpeg-sys-next-80de3680ed4e4b8b\build-script-build` (exit code: 101)

H:\gifski-main>
```

- I have LLVM_clang 14.0.6 installed and path is set in the environment
- Tried with the files from the link above for FFMPEG, and also from the package i compiled using VS2022 and the vcpkg package version 4.4.1#13 we can find on vcpkg repo
- I set the FFMPEG path to either one of those and always end up with an error when launching the _cargo run --features=video-prebuilt-static_ command from Gifski folder.

Am i doing it wrong or missing something please ? 

I'd rather use the files i have from vcpkg as I enabled the [Avisynth+](https://github.com/AviSynth/AviSynthPlus) switch there so i could feed Gifski directly with _.avs_ files.

Right now i already have [a simplified way](https://github.com/jalba69/Staxrip_Stuff/tree/master/Gif_with_Gifski) to use Gifski but if i could avoid the step where i need to extract the png's it will be a big plus, no matter if i need the dll's from ffmpeg.



PS : also saw  _vcpkg-rs_ with the [_RUSTFLAGS=-Ctarget-feature=+crt-static_](https://docs.rs/vcpkg/latest/vcpkg/) parameter that could maybe help avoid the separate dll's, well i think... this is far behind my knowledge and after looking around i still have no clue how to use this in a rust package. 
Anyways maybe something to explore for coding guru's, will make things way easier, at least for me who is used to getting vcpkg lib's with visual studio and already have the ffmpeg ones.

Thanks a lot.

**Comment by jalba69** [comment]
Should have waited to post my previous comment...

after many more failed attempts i have it finally working, seems like adding _LIBCLANG_PATH_ to env. variable did the trick, i already had the rest correctly set and according to the doc it's one or the other if you have LLVM path set but I'm just haPpy it works.

Used the FFmpeg files from [windows vcpkg](https://github.com/microsoft/vcpkg) package i had already, of course i still need FFmpeg dll's like mentioned in your attempts but this is just a minor annoyance compared to the benefits of having FFmpeg built-in.

Bonus : it have avisynth support, so i can feed Gifski directly with a simple .avs script containing a few lines to trim,crop,resize video using a better resizing filter or add any other avisynth supported filters/plugins to it in a blink which is what i had in mind since a discovered this tool some years ago.


Now it is time to play with this and test it deeply, thank you very much guys :)

**Comment by jalba69** [comment]
So... 
I have it now compiled as a single .exe with ffmpeg support and no dll's needed with it :)

getting the ffmpeg static files needed : 
`vcpkg install ffmpeg:x64-windows-static`

then :
`FFMPEG_DIR=`   set to    `\vcpkg\installed\x64-windows-static` folder

and most important to fix the link.exe errors i had because of how vcpkg works and skip libraries : 
- create a `config.toml` file with this inside :
```
[target.x86_64-pc-windows-msvc]
rustflags = [
    "-C", "link-arg=dxva2.lib",
    "-C", "link-arg=evr.lib",
    "-C", "link-arg=mf.lib",
    "-C", "link-arg=mfplat.lib",
    "-C", "link-arg=mfplay.lib",
    "-C", "link-arg=mfreadwrite.lib",
    "-C", "link-arg=mfuuid.lib",
    "-C", "link-arg=bcrypt.lib",
    "-C", "link-arg=ws2_32.lib",
    "-C", "link-arg=Secur32.lib",
    "-C", "link-arg=Strmiids.lib",
    "-C", "link-arg=Ole32.lib"
]
```
and put this file inside the .cargo main folder 

Then `cargo build --release --features=video-prebuilt-static`  in Gifski main folder and i end up with a 18.9MB .exe and it works for the few tries i done ;)

Only big down for me is that vcpkg doesn't seems able to get the static FFMEG with avisynth feature enabled, seems possible only with a dynamic version, but i'll have to see this with ffmpeg's vcpkg guys directly in the next days I guess, hopefully this can be fixed since we can have a avisynth static version from vcpkg.

Anyways, after spending so many time on this it works so I'm happy and it might help someone else who knows ^^

**Comment by megapro17** [comment]
@jalba69 upload EXE

**Comment by jalba69** [comment]
Here it is :

[gifski.zip](https://github.com/ImageOptim/gifski/files/9832957/gifski.zip) v1.7.2.0

SHA-256 for the zipped file : _AB05114DBF0CEAF9974A0C1D1A2346841726A334E4D387E8C7697E549FB6142C_
SHA-256  for the .exe once unzipped : _7AC8F5A8A28BF39D26C9F1B398651F92EA081767A4CD0921259D2674DA8843FF_

VirusTotal scan for the zip : https://www.virustotal.com/gui/url-analysis/u-2dbffbefbbc162183294b2a1f6ae021d28fbc48647cda708e92117900967185e-1666291294

It have _ffmpeg & avisynth_ support + added infos details in the proprieties  and a dedicated icon for the .exe using a .res file.
It also use the _mimalloc 0.1.29_ crate that give slightly better performances for me on windows.

Enjoy ;)

@kornelski : if you have a problem with me uploading the .exe here let me know, i'll remove my post asap, thanks again for this nice tool :)

**Comment by megapro17** [comment]
Thanks, but why you're worrying so much..

**Comment by jalba69** [comment]
not worried, just used to do it that way, plus this might start to travel around on the net so adding infos from the starting source ;)

**Comment by God-damnit-all** [comment]
Never did manage to figure this out sadly.
